File: graph/29.py
import heapq
from typing import List
import math


class Edge:
    def __init__(self, w: int, weight: int, time: int):
        self.w = w
        self.weight = weight
        self.time = time

    def __lt__(self, other):
        return self.weight > other.weight


def dijkstra(
    graph: List[List[Edge]],
    source: int,
    dest: int,
    stops: int,
    dist: List,
    is_answer: List,
):
    heap = []
    dist[source] = 0
    heapq.heappush(heap, Edge(source, 0, 0))

    while len(heap) > 0:
        edge = heapq.heappop(heap)

        if edge.w == dest:
            is_answer[0] = True
            return

        for v in graph[edge.w]:
            if dist[v.w] > dist[edge.w] + v.weight and edge.time <= stops:
                dist[v.w] = dist[edge.w] + v.weight
                heapq.heappush(heap, Edge(v.w, dist[v.w], edge.time + 1))

    return -1


def findCheapestPrice(
    n: int, flights: List[List[int]], src: int, dst: int, k: int
) -> int:
    dist = [math.inf] * n
    graph = [[] for _ in range(n)]
    is_answer = [False]

    for f in flights:
        graph[f[0]].append(Edge(f[1], f[2], 0))

    dijkstra(graph, src, dst, k, dist, is_answer)

    if is_answer[0] is True:
        return dist[dst]

    return -1


flights = [
    [0, 1, 100],
    [0, 2, 100],
    [2, 3, 1],
    [1, 4, 602],
    [3, 4, 600],
    [4, 5, 6],
    [5, 6, 6],
]
n = 7
src = 0
dst = 6
k = 3

print(findCheapestPrice(n, flights, src, dst, k))

File: graph/29.cpp
#include <bits/stdc++.h>

using namespace std;

template <typename T>
using Graph = vector<vector<T>>;

struct edge {
    int w;
    int weight;
    int time;

    bool operator<(const edge &other) const {
        return weight < other.weight;
    }
};

void dijkstra(Graph<edge> &G, int source, int dest, int stops, bool &isAnswer, vector<int> &dist) {
    priority_queue<edge, vector<edge>> q;
    dist[source] = 0;

    q.push({source, 0, 0});

    while (q.size() > 0) {
        auto item = q.top();
        q.pop();

        if (item.w == dest) {
            isAnswer = true;
            return;
        }

        for (const auto &u : G[item.w]) {
            if (dist[u.w] > dist[item.w] + u.weight && item.time <= stops) {
                dist[u.w] = dist[item.w] + u.weight;
                q.push({u.w, dist[u.w], item.time + 1});
            }
        }
    }
}

int findCheapestPrice(int n, vector<vector<int>> &flights, int src, int dst, int k) {
    Graph<edge> G(n);
    vector<int> dist(n, INT_MAX);
    bool isAnswer = false;

    for (const auto &v : flights) {
        G[v[0]].push_back({v[1], v[2], 0});
    }

    dijkstra(G, src, dst, k, isAnswer, dist);

    if (isAnswer)
        return dist[dst];

    return -1;
}

int main() {
    vector<vector<int>> flights = {
        {0, 1, 100}, 
        {0, 2, 100}, 
        {2, 3, 1}, 
        {3, 4, 600}, 
        {1, 4, 602}, 
        {4, 5, 6}, 
        {5, 6, 6}
    };

    int src = 0;
    int dst = 6;
    int k = 3;
    int n = 7;

    int ans = findCheapestPrice(n, flights, src, dst, k);
    cout << ans;

    return 0;
}